In [99]:
# Imports and pip installations (if needed)
!python -m venv sklearn-env
!source sklearn-env/bin/activate
%pip install -U scikit-learn
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier


Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [100]:
# Load the dataset (load remotely or locally)
df = pd.read_csv('iris.csv')
print(df.head(15))
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
summary_stats = df.describe()
print(summary_stats)
print(df.dtypes)


    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

* sepal_length: The length of the flower's sepal. (feature)

* sepal_width: The width of the flower's sepal. (feature)

* petal_length: The length of the pedal. (feature)

* petal_width: The width of the petal. (feature)

* species: The species of the flower. (label)

The mapping of labels:

0: setosa

1: versicolor

2: virginica


# Part 2: Split the dataset into train and test

In [101]:
# Take the dataset and split it into our features (X) and label (y)
X = df.loc[:, df.columns != 'species']
y = df['species']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Part 3: Logistic Regression

In [102]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression(max_iter=200)
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_index = 0  # Change this index to select a different sample
sample_data = X_test.iloc[sample_index].values.reshape(1, -1)  # Reshape it into a 2D array
# Convert back to DataFrame with feature names
sample_data_df = pd.DataFrame(sample_data, columns=X_test.columns)
probabilities = model.predict_proba(sample_data_df)
print(f'Sample data: {sample_data}')
print(f'Predicted probabilities for the sample data point: {probabilities}')

predicted_class = np.argmax(probabilities, axis=1)
print(f'Predicted Class: {predicted_class}')
class_labels = model.classes_  # This will give you the class labels
print("Class Labels:", class_labels)
predicted_class_name = class_labels[predicted_class[0]]
print(f'Predicted Class (Name): {predicted_class_name}')

# iii. Report on the score for Logistic regression model, what does the score measure?
scores = cross_val_score(model, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')

"""
    Report: The score of accuracy on the test data is: 0.9625541125541126
"""

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

Sample data: [[6.1 2.8 4.7 1.2]]
Predicted probabilities for the sample data point: [[0.00345024 0.8365213  0.16002846]]
Predicted Class: [1]
Class Labels: ['setosa' 'versicolor' 'virginica']
Predicted Class (Name): versicolor


Cross-Validation Scores: [1.         1.         1.         1.         0.85714286 0.92307692
 1.         1.         1.         0.92307692]
Mean Cross-Validation Accuracy: 0.9703296703296704
Coefficients: [[-0.42619161  0.96870363 -2.44814477 -1.03770553]
 [ 0.51154842 -0.21828724 -0.21576926 -0.84253553]
 [-0.0853568  -0.75041638  2.66391403  1.88024106]]
Intercept: [  9.52968696   1.89748664 -11.42717361]


# Part 4: Support Vector Machine

In [103]:
# i. Use sklearn to train a Support Vector Classifier on the training set
model = SVC(kernel='linear', probability=True)  # Use a linear kernel; 'probability=True' allows for probability estimates
model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_index = 0  # Change this index to select a different sample
sample_data = X_test.iloc[sample_index].values.reshape(1, -1)  # Reshape it into a 2D array
# Convert back to DataFrame with feature names
sample_data_df = pd.DataFrame(sample_data, columns=X_test.columns)
probabilities = model.predict_proba(sample_data_df)
print(f'Sample data: {sample_data}')
print(f'Predicted probabilities for the sample data point: {probabilities}')

predicted_class = np.argmax(probabilities, axis=1)
print(f'Predicted Class: {predicted_class}')
class_labels = model.classes_  # This will give you the class labels
print("Class Labels:", class_labels)
predicted_class_name = class_labels[predicted_class[0]]
print(f'Predicted Class (Name): {predicted_class_name}')

# iii. Report on the score for the SVM, what does the score measure?
# Perform 10-fold cross-validation to evaluate the model
scores = cross_val_score(model, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')

"""
    Report: The score of accuracy on the test data is: 0.9718614718614719
"""

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

Sample data: [[6.1 2.8 4.7 1.2]]
Predicted probabilities for the sample data point: [[0.00589737 0.9149765  0.07912613]]
Predicted Class: [1]
Class Labels: ['setosa' 'versicolor' 'virginica']
Predicted Class (Name): versicolor
Cross-Validation Scores: [1.         1.         1.         1.         1.         0.84615385
 1.         1.         1.         1.        ]
Mean Cross-Validation Accuracy: 0.9846153846153847
Coefficients: [[-0.04631136  0.52105578 -1.0030165  -0.46411816]
 [-0.00641373  0.17867392 -0.5389119  -0.29158729]
 [ 0.6136048   1.13849257 -2.00914998 -1.78734049]]
Intercept: [1.45338269 1.50537328 5.67958424]


# Part 5: Neural Network

In [104]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=1000, random_state=42)  # You can customize the architecture
mlp_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_index = 0  # Change this index to select a different sample
sample_data = X_test.iloc[sample_index].values.reshape(1, -1)  # Reshape it into a 2D array
# Convert back to DataFrame with feature names
sample_data_df = pd.DataFrame(sample_data, columns=X_test.columns)
probabilities = mlp_model.predict_proba(sample_data_df)
print(f'Sample data: {sample_data}')
print(f'Predicted probabilities for the sample data point: {probabilities}')

predicted_class = np.argmax(probabilities, axis=1)
print(f'Predicted Class: {predicted_class}')
class_labels = mlp_model.classes_  # This will give you the class labels
print("Class Labels:", class_labels)
predicted_class_name = class_labels[predicted_class[0]]
print(f'Predicted Class (Name): {predicted_class_name}')

# iii. Report on the score for the Neural Network, what does the score measure?
# Perform 10-fold cross-validation to evaluate the model
scores = cross_val_score(mlp_model, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')

"""
    Report: The score of accuracy on the test data is: 0.9764069264069264
"""

# iv: Experiment with different options for the neural network, report on your best configuration
print()

# Experiment 1: Default Settings
mlp_default = MLPClassifier(max_iter=1000, random_state=42)
mlp_default.fit(X_train, y_train)
y_pred_default = mlp_default.predict(X_test)
print("Experiment 1: Default Settings")
scores = cross_val_score(mlp_default, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 2: Increase the number of hidden layers and neurons
mlp_layers = MLPClassifier(hidden_layer_sizes=(100, 50, 25), max_iter=1000, random_state=42)
mlp_layers.fit(X_train, y_train)
y_pred_layers = mlp_layers.predict(X_test)
print("Experiment 2: Increased Hidden Layers")
scores = cross_val_score(mlp_layers, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 3: Change the activation function
mlp_tanh = MLPClassifier(hidden_layer_sizes=(100, 50), activation='tanh', max_iter=1000, random_state=42)
mlp_tanh.fit(X_train, y_train)
y_pred_tanh = mlp_tanh.predict(X_test)
print("Experiment 3: Tanh Activation Function")
scores = cross_val_score(mlp_tanh, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 4: Change the learning rate initialization
mlp_lr = MLPClassifier(hidden_layer_sizes=(100, 50), learning_rate_init=0.01, max_iter=1000, random_state=42)
mlp_lr.fit(X_train, y_train)
y_pred_lr = mlp_lr.predict(X_test)
print("Experiment 4: Increased Learning Rate Initialization")
scores = cross_val_score(mlp_lr, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 5: Add L2 regularization
mlp_alpha = MLPClassifier(hidden_layer_sizes=(100, 50), alpha=0.1, max_iter=1000, random_state=42)
mlp_alpha.fit(X_train, y_train)
y_pred_alpha = mlp_alpha.predict(X_test)
print("Experiment 5: L2 Regularization")
scores = cross_val_score(mlp_alpha, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 6: Use Early Stopping
mlp_early_stopping = MLPClassifier(hidden_layer_sizes=(100, 50), max_iter=1000, early_stopping=True, random_state=42)
mlp_early_stopping.fit(X_train, y_train)
y_pred_early_stopping = mlp_early_stopping.predict(X_test)
print("Experiment 6: Early Stopping")
scores = cross_val_score(mlp_early_stopping, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 7: Use a different Solver (e.g. 'adam')
mlp_solver = MLPClassifier(hidden_layer_sizes=(100, 50), solver='adam', max_iter=1000, random_state=42)
mlp_solver.fit(X_train, y_train)
y_pred_solver = mlp_solver.predict(X_test)
print("Experiment 7: Using 'adam' Solver")
scores = cross_val_score(mlp_solver, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')
print()

# Experiment 8: Adaptive learning rate
mlp_adaptive = MLPClassifier(hidden_layer_sizes=(100, 50), learning_rate='adaptive', max_iter=1000, random_state=42)
mlp_adaptive.fit(X_train, y_train)
y_pred_adaptive = mlp_adaptive.predict(X_test)
print("Experiment 8: Using adaptive learning rate")
scores = cross_val_score(mlp_adaptive, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')

"""
    Report: experiment 1 using the default setting and no hidden layer has the best performance out of all
    8 experiments, with an accuracy of 0.9811688311688311.
"""

Sample data: [[6.1 2.8 4.7 1.2]]
Predicted probabilities for the sample data point: [[3.05886003e-04 9.92594953e-01 7.09916127e-03]]
Predicted Class: [1]
Class Labels: ['setosa' 'versicolor' 'virginica']
Predicted Class (Name): versicolor
Cross-Validation Scores: [1.         1.         0.92857143 1.         0.92857143 0.84615385
 1.         1.         1.         1.        ]
Mean Cross-Validation Accuracy: 0.9703296703296704

Experiment 1: Default Settings
Cross-Validation Scores: [1.         1.         0.92857143 1.         0.92857143 0.84615385
 1.         1.         1.         1.        ]
Mean Cross-Validation Accuracy: 0.9703296703296704

Experiment 2: Increased Hidden Layers
Cross-Validation Scores: [1.         1.         0.92857143 1.         0.92857143 0.84615385
 1.         1.         1.         1.        ]
Mean Cross-Validation Accuracy: 0.9703296703296704

Experiment 3: Tanh Activation Function
Cross-Validation Scores: [1.         1.         0.92857143 1.         0.92857143 0.

'\n    Report: experiment 1 using the default setting and no hidden layer has the best performance out of all\n    8 experiments, with an accuracy of 0.9811688311688311.\n'

# Part 6: K-Nearest Neighbors

In [105]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn_model = KNeighborsClassifier(n_neighbors=3)  # You can adjust n_neighbors
knn_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_index = 0  # Change this index to select a different sample
sample_data = X_test.iloc[sample_index].values.reshape(1, -1)  # Reshape it into a 2D array
# Convert back to DataFrame with feature names
sample_data_df = pd.DataFrame(sample_data, columns=X_test.columns)
probabilities = knn_model.predict_proba(sample_data_df)
print(f'Sample data: {sample_data}')
print(f'Predicted probabilities for the sample data point: {probabilities}')

predicted_class = np.argmax(probabilities, axis=1)
print(f'Predicted Class: {predicted_class}')
class_labels = knn_model.classes_  # This will give you the class labels
print("Class Labels:", class_labels)
predicted_class_name = class_labels[predicted_class[0]]
print(f'Predicted Class (Name): {predicted_class_name}')

# iii. Report on the score for kNN, what does the score measure?
# Perform 10-fold cross-validation to evaluate the model
scores = cross_val_score(knn_model, X_train, y_train, cv=10)
# Output the cross-validation scores and mean accuracy
print("Cross-Validation Scores:", scores)
print(f'Mean Cross-Validation Accuracy: {np.mean(scores)}')

"""
    Report: The score of accuracy on the test data is: 0.9623376623376624
"""


Sample data: [[6.1 2.8 4.7 1.2]]
Predicted probabilities for the sample data point: [[0. 1. 0.]]
Predicted Class: [1]
Class Labels: ['setosa' 'versicolor' 'virginica']
Predicted Class (Name): versicolor
Cross-Validation Scores: [1.         1.         0.92857143 1.         0.92857143 0.84615385
 1.         1.         1.         0.92307692]
Mean Cross-Validation Accuracy: 0.9626373626373628


'\nReport: The score of accuracy on the test data is: 0.9623376623376624\n'

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

A: This notebook trains a model that takes in a flower's sepal length, speal width, petal length and petal width to classify it into one of the following species: setosa, versicolor, virginica. The result of all models have an accuracy above 95%, with the MLP Classifier having the best score above 98%. I think that is because the size of the dataset is too small and MLP could be potentially overfitting the test data. One thing I was surprised with is that among all the MLP Classifiers I attempted, the default one (with no hidden layer) got the best score.